core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -590,56 +648,46 @@ public CreatePathResult createPath(RpcContext rpcContext, LockedInodePath inodeP
             currentInodeDirectory, pathComponents[k]);
 
-        if (!currentInodeDirectory.addChild(dir)) {
-          // The child directory inode already exists. Get the existing child inode.
-          extensibleInodePath.getLockList().unlockLast();
+        newDir.setPinned(currentInodeDirectory.isPinned());
 
-          dir =
-              (InodeDirectory) currentInodeDirectory.getChildReadLock(pathComponents[k],
-                  extensibleInodePath.getLockList());
-          if (dir == null) {
-            // Could not get the child inode. Continue and try again.
-            continue;
-          }
-        } else {
-          try {
-            // Successfully added the child, while holding the write lock.
-            dir.setPinned(currentInodeDirectory.isPinned());
-
-            // if the parent has default ACL, copy that default ACL as the new directory's default
-            // and access acl.
-            if (!options.isMetadataLoad()) {
-              DefaultAccessControlList dAcl = currentInodeDirectory.getDefaultACL();
-              if (!dAcl.isEmpty()) {
-                Pair<AccessControlList, DefaultAccessControlList> pair = dAcl.generateChildDirACL();
-                dir.setInternalAcl(pair.getFirst());
-                dir.setDefaultACL(pair.getSecond());
-              }
-            }
-            if (options.isPersisted()) {
-              // Do not journal the persist entry, since a creation entry will be journaled instead.
-              syncPersistDirectory(RpcContext.NOOP, dir);
-            }
-          } catch (Exception e) {
-            // Failed to persist the directory, so remove it from the parent.
-            currentInodeDirectory.removeChild(dir);
-            throw e;
+        // if the parent has default ACL, copy that default ACL as the new directory's default
+        // and access acl.
+        if (!options.isMetadataLoad()) {
+          DefaultAccessControlList dAcl = currentInodeDirectory.getDefaultACL();
+          if (!dAcl.isEmpty()) {
+            Pair<AccessControlList, DefaultAccessControlList> pair = dAcl.generateChildDirACL();
+            newDir.setInternalAcl(pair.getFirst());
+            newDir.setDefaultACL(pair.getSecond());
           }
-          // Journal the new inode.
-          rpcContext.getJournalContext().append(dir.toJournalEntry());
-          mInodes.add(dir);
+        }
+        if (options.isPersisted()) {
+          syncPersistNewDirectory(newDir);
+        }
 
+        if (mState.applyAndJournal(rpcContext, newDir)) {
           // After creation and journaling, downgrade to a read lock.
           extensibleInodePath.getLockList().downgradeLast();
+          dir = newDir;
+        } else {
+          // The child directory inode already exists. Get the existing child inode.
+          extensibleInodePath.getLockList().unlockLast();
+
+          InodeView existing = currentInodeDirectory.getChildReadLock(pathComponents[k],
```

Comment:
shouldn't the old behavior be copied? If the node was added then removed, this getChild could return null, and we should continue with the loop?

Reply:
great catch, fixed. I'll see if I can add a test that would catch this